app/models/concerns/thredded/topic_common.rb

```diff
@@ -47,22 +47,30 @@ def unread(user)
           .merge(reads_class.where(reads[:id].eq(nil).or(reads[:read_at].lt(topics[:updated_at]))))
       end
 
+      def read_states_by_topics_lookup(user)
+        read_states_by_topic_id =
+          reflect_on_association(:user_read_states).klass
+            .where(user_id: user.id, postable_id: current_scope.map(&:id))
+            .group_by(&:postable_id)
+
+        def read_states_by_topic_id.get(topic, null_value = nil)
```

Comment:
This flushes MRI method cache on every call to the parent method, which is very slow. Seems that you can simply use a Hash with a default block here. http://ruby-doc.org/core-2.2.0/Hash.html#method-c-new